File: dart/conformance/test/optional_identity_test.dart
// Copyright 2026 Avala AI
// SPDX-License-Identifier: Apache-2.0

/// End-to-end checks for the shared optional-identity witnesses.
library;

import 'dart:convert';
import 'dart:io';

import 'package:fourdgs_conformance/canonical.dart';
import 'package:test/test.dart';

const Map<String, String> _witnesses = <String, String>{
  'gaussian-birth':
      'identity/gaussian-birth/'
      'OptionalIdentityGaussianBirth-UseChunkIndex-UseCrc',
  'keyframe-delta':
      'identity/keyframe-delta/'
      'OptionalIdentityKeyframeDelta-UseChunkIndex-UseCrc-UseStatistics',
};

void main() {
  for (final runner in <String>[
    'decode_streamed.dart',
    'decode_indexed.dart',
  ]) {
    for (final witness in _witnesses.entries) {
      test('$runner reconstructs ${witness.key} optional identities', () {
        final stem = '../../tests/conformance/data/${witness.value}';
        final input = File('$stem.4dgs');
        final expectation = File('$stem.json');
        expect(input.existsSync(), isTrue, reason: 'generate the corpus first');
        expect(
          expectation.existsSync(),
          isTrue,
          reason: 'the witness needs its committed expectation',
        );

        final done = Process.runSync(Platform.resolvedExecutable, <String>[
          'run',
          'bin/$runner',
          input.path,
        ]);
        expect(done.exitCode, 0, reason: done.stderr as String);
        expect(
          canonical(jsonDecode(done.stdout as String)),
          canonical(jsonDecode(expectation.readAsStringSync())),
        );
      });
    }
  }
}

File: typescript/conformance/src/optionalIdentity.test.ts
// Copyright 2026 Avala AI
// SPDX-License-Identifier: Apache-2.0

/** Focused integration tests for the optional-identity shared witnesses. */

import assert from "node:assert/strict";
import { existsSync, readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { test } from "node:test";

import {
  BytesReadable,
  IndexedDecoder,
  assembleGaussians,
  decodeKeyframeDeltaIndexed,
  decodeKeyframeDeltaStreamed,
  decodeScene,
  type ChunkGaussians,
} from "@4dgs/core";

import { canonical } from "./canonical.js";
import { gaussianBirthIdentityJson, keyframeDeltaIdentityJson } from "./optionalIdentity.js";

const DATA = fileURLToPath(new URL("../../../tests/conformance/data/identity/", import.meta.url));

function witness(relative: string): { bytes: Uint8Array; expected: string } | null {
  const stem = `${DATA}${relative}`;
  if (!existsSync(`${stem}.4dgs`)) return null;
  return {
    bytes: new Uint8Array(readFileSync(`${stem}.4dgs`)),
    expected: canonical(JSON.parse(readFileSync(`${stem}.json`, "utf8"))),
  };
}

test("gaussian-birth mixed chunks zero-fill all optional identities on both paths", async (t) => {
  const fixture = witness("gaussian-birth/OptionalIdentityGaussianBirth-UseChunkIndex-UseCrc");
  if (fixture === null) return t.skip("corpus not generated");

  const streamed = await decodeScene(fixture.bytes);
  assert.equal(canonical(gaussianBirthIdentityJson(streamed.gaussians)), fixture.expected);

  const indexed = await IndexedDecoder.open(new BytesReadable(fixture.bytes));
  const chunks: ChunkGaussians[] = [];
  for (const entry of indexed.index) chunks.push((await indexed.readChunk(entry)).gaussians);
  const gaussians = assembleGaussians(chunks, indexed.windows, indexed.header.shDegree);
  assert.equal(canonical(gaussianBirthIdentityJson(gaussians)), fixture.expected);
  assert.deepEqual([...gaussians.sourceGroup!].slice(-2), [0, 0]);
  assert.deepEqual([...gaussians.sourceIndex!].slice(-2), [0, 0]);
  assert.deepEqual([...gaussians.objectId!], [0x80000000, 0xffffffff, 0, 0, 0]);
});

test("keyframe-delta composes all optional identity transitions on both paths", async (t) => {
  const fixture = witness(
    "keyframe-delta/OptionalIdentityKeyframeDelta-UseChunkIndex-UseCrc-UseStatistics",
  );
  if (fixture === null) return t.skip("corpus not generated");

  const streamed = await decodeKeyframeDeltaStreamed(fixture.bytes);
  const indexed = (await decodeKeyframeDeltaIndexed(fixture.bytes)).sequence;
  for (const sequence of [streamed, indexed]) {
    assert.equal(canonical(keyframeDeltaIdentityJson(sequence)), fixture.expected);

    const introduced = sequence.chunks[1]!.state;
    assert.deepEqual(introduced.identityAt(0), {
      sourceGroup: -17,
      sourceIndex: 23,
      objectId: 0xffffffff,
    });
    assert.deepEqual(introduced.identityAt(1), {
      sourceGroup: 0,
      sourceIndex: 0,
      objectId: 0,
    });

    const reset = sequence.chunks[4]!.state;
    for (let row = 0; row < reset.count; row++) {
      assert.deepEqual(reset.identityAt(row), { sourceGroup: 0, sourceIndex: 0, objectId: 0 });
    }
  }
});
